Replacing OIS with SDL input (using OGRE14 + Bites component)

ohlidalp opened the pull request:
This is a first step in migrating from self-shipped OIS library to standard SDL2 and OGRE's builtin DearIMGUI integration. I've based it on our (pretty mature already) OGRE13 port to get the latest tech and support.

**Only mouse+keyboard are working, controllers and joysticks (+ force feedback) are temporarily disabled.**

OGRE provides Bites component with generic ApplicationContext that runs on many platforms and powers OGRE SampleBrowser. It includes DearIMGUI integration, RTSS setup, window creation and generic input API for all devices. I tried replacing our OIS events + API calls 1:1 with Bites events and SDL API calls and it turns out to work just fine. I had to also replace the function which opens render window, as SDL input events don't work otherwise. Everything else remains unchanged. Also our (very outdated) DearIMGUI integration remains in place, with only updated input processing (based on OgreBites code, anyway).

ohlidalp committed:
port to Ogre 13 (without RTSS)

Joint effort by @ohlidalp, @Tritonas00, @AnotherFoxGuy, @CuriousMike56

:arrow_up: Updated mygui, pagedgeometry and caelum
:bug: Fixed missing MyGUI shaders
:bug: Fixed MyGUI rendering issue
:wrench: Build OGRE with resourcemanager_strict=off
updated rtshader media
:bug: Fixed cmake not accepting `CMAKE_CONFIGURATION_TYPES`
fixed odef lights scenenode
:bug: Fixed Caelum not loading os files
Caelum time slider: restored original code (it works with latest Ogre master branch)
:bug: Fixed Imgui inverted colours on DirectX 9
fixed survey map texture updating
:arrow_up: Update OGRE to 13.5
updated rtshader media for Ogre 13.5
:wrench: Implement PSSM3 properly
:arrow_up: Update OGRE to 13.5.1
Fixed crash on loading 2nd terrain w/envmap enabled
:arrow_up: Update OGRE to 13.5.3
Fixed crash on 2nd map load (procedural road material conflict)
Media: fix corrupted DDS texture crashing under D3D9

ohlidalp committed:
Replaced OIS with OgreBites input + SDL (not functional yet)

Status: builds and runs OK, but there is no input.

This is a brainless 1:1 replacement of OIS API and events with SDL API and OgreBites input events. I preserved our (very outdated) DearIMGUI integration. All our (quite convoluted) input processing stayed mostly intact. Only mouse/keyboard input is handled, joysticks + controllers + forcefeedback are temporarily disabled.

ohlidalp committed:
Working SDL input (keyboard/mouse)

Status: build and runs, keyboard+mouse working, window is somehow oversized but responsive.

Codechanges:
* creating render window via OgreBites::ApplicationContext instead of OGRE Root directly - needed to receive events
* overriding OgreBites::ApplicationContext::createRoot() with our pre-existing setup code, to initialize `mRoot` as expected.
* calling `OgreBites::ApplicationContext::addInputListener()` to receive events.

ohlidalp committed:
OGRE13: attached beacon lights to scenenodes

only-a-ptr commented:
I tried replacing our DearIMGUI integration (v 1.75 - very outdated) with OGRE's builtin DearImguiOverlay component. I made a lot of small commits to keep a track record, I'll squash it to one commented commit when ready.

Status:
* game builds and runs
* UI draws correctly except it uses the builtin font
* Keyboard input works for both simulation and UI, mouse is dead completely.
* Under Debug I get asserts about `NewFrame()/EndFrame()` mismatches.

ohlidalp committed:
Migrated to OGRE's built-in DearIMGUI

STATUS:
- ogre imgui port: keyboard/mouse input working!
- ogre imgui port: FIXME: hacked Renderdash to build, but imgui will likely render into it - how to disable??
- ogre imgui port: builds and runs with default font.
- under Debug I get asserts about NewFrame()/EndFrame() mismatches, to be fixed later.

CODECHANGES:
* imgui deprecated - renamed ImGui::AlignFirstTextHeightToWidgets() to ImGui::AlignTextToFramePadding()
* imgui deprecated - ImGui::GetContentRegionAvailWidth() -> use ImGui::GetContentRegionAvail().x
* imgui deprecated -  GetWindowContentRegionWidth() function. -> 'GetWindowContentRegionMax().x
* imgui deprecated - SetScrollHere() -> SetScrollHereX() + SetScollHereY()
* imgui deprecated - ImGui::TreeAdvanceToLabelPos()      -> use ImGui::SetCursorPosX(ImGui::GetCursorPosX() + ImGui::GetTreeNodeToLabelSpacing());
* imgui deprecated - renamed SetNextTreeNodeOpen() to SetNextItemOpen().
* imgui deprecated - renamed ListBoxHeader(const char* label, ImVec2 size) to BeginListBox().
* imgui deprecated - removed BeginPopupContextWindow(const char*, int mouse_button, bool also_over_items) in favor of BeginPopupContextWindow(const char*, ImGuiPopupFlags flags) with ImGuiPopupFlags_NoOverItems.
* imgui deprecated - removed CalcListClipping() function. Prefer using ImGuiListClipper which can return non-contiguous ranges.
* imgui deprecated - ImGuiCol_ChildWindowBg              -> use ImGuiCol_ChildBg
* imgui deprecated - style: renamed ImGuiCol_ModalWindowDarkening to ImGuiCol_ModalWindowDimBg
* imgui deprecated - GetItemsLineHeightWithSpacing()     -> use GetFrameHeightWithSpacing()
* imgui deprecated - InputText: Renamed  ImGuiTextEditCallbackData to ImGuiInputTextCallbackData
* imgui deprecated - SetNextWindowContentWidth(float w)    ->    SetNextWindowContentSize(ImVec2(w, 0.0f))
* imgui deprecated - SetNextWindowPosCenter(ImGuiCond c=0) { SetNextWindowPos(GetMainViewport()->GetCenter(), c, ImVec2(0.5f,0.5f));
* ogre imgui port: use public `GetFont()/GetFontSize()` instead of internal `ImGuiContext->Font/FontSize`
* VehicleAI panel: `Push/PopItemFlag(Disabled)` (imgui_internal) -> `Begin/EndDisabled()` (public BETA API)

ohlidalp committed:
OGRE13 Build fix - missing includes in GenericFileFormat.h

ohlidalp committed:
OGRE13-SDL: Fixed window appearing on top left with titlebar outside of screen.

Deleted the windows-specific code setting ["monitorIndex"]- apparently OGRE does it by itself in `ApplicationContextSDL::createWindow()`, see OgreApplicationContextSDL.cpp line 56.
Original commit: 128a8e269230da7c210f16d5db1ea513f71cde92 - "Attempt to fix the 'Rendering Device' setting" - Ulteq, 4 years ago (29.12.2018 13:53:57)

Also deleted setting ["windowProc"] - doesn't seem to have any effect, perhaps an OIS leftover.

ohlidalp committed:
OGRE13 build fix - undefined class Actor in ProjectedGrid.cpp

ohlidalp committed:
OGRE13 imgui: fixed interference with renderdash

ohlidalp committed:
OGRE13 imgui: Re-added font

ohlidalp committed:
Debug build fix of #3085

ohlidalp committed:
OGRE14 migration (WIP, build broken)

Codechanges:
* GUIUtils, InputEngine: `getModifierKeyName()` replaced OIS keycodes with OgreBites/SDL keycodes
* ImGuiAngelscript: commented out some dead items, added (ImGuiKey) casts where needed, see https://github.com/ocornut/imgui/issues/4921
* GUI_VehicleButtons: build fix for `ImGui::ImageButton()` - added parameter ID-string.
* GUI_RepositorySelector: updated thumbnail downloading to use new OGRE WorkQueue API, see https://github.com/OGRECave/ogre/blob/master/Docs/14-Notes.md#task-based-workqueue
* Application.h, main.cpp: added :envelope: `MSG_NET_DOWNLOAD_REPOTHUMB_SUCCESS`, used instead of OGRE WorkQueue's `addMainThreadTask` for consistency across the various background threads in the codebase.
* GodRaysManager: build fix - RENDER_QUEUE_9 was removed from OGRE, using 8.
* RttManager: rendersystem-specific `getProjectionMatrixRS()` was removed from OGRE, using portable  `getProjectionMatrix()`
